detangle_client_base_classes.rst

```diff
@@ -0,0 +1,224 @@
+De-tangling Client Base Classes
+===============================
+
+Survey client classes
+---------------------
+
+Figure out which client classes consume which base classes.
+
+
+Autogen clients
+~~~~~~~~~~~~~~~
+
+These clients are auto-genereated, and do all API requests over gRPC:
+
+- ``asset``
+- ``automl``
+- ``bigquery_datatransfer``
+- ``container``
+- ``dataproc``
+- ``dlp``
+- ``iot``
+- ``kms``
+- ``language``
+- ``monitoring`` (handwritten ``query`` module uses only one client method).
+- ``oslogin``
+- ``redis``
+- ``speech`` (client mixes in ``SpeechHelpers`` oddly)
+- ``tasks``
+- ``texttospeech``
+- ``trace``
+- ``videointelligence``
+- ``vision``
+- ``websecurityscanner``
+
+None of them import from ``google.cloud.client``.
+
+
+Manual GAPIC-only clients
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+These clients do all API requests over gRPC:
+
+- ``bigtable``
+- ``firestore``
+- ``pubsub`` (``PublisherClient`` and ``SubscriberClient``).
+- ``spanner``
+
+Of these, the following still import from ``google.cloud.client``:
+
+- ``bigtable``
+- ``firestore``
+- ``spanner``
+
+
+HTTP-only clients
+~~~~~~~~~~~~~~~~~
+
+These clients do all API requests over HTTP:
+
+- ``bigquery``
+- ``dns``
+- ``resource_manager``
+- ``runtimeconfig``
+- ``storage``
+- ``translate``
+
+
+Hybrid clients
+~~~~~~~~~~~~~~
+
+These clients allow selecting between gRPC and HTTP via an environment
+variable, used to select which API wrapper to construct.
+
+- ``datastore``
+- ``error_reporting``
+- ``logging``
+
+
+Base Classes Defined
+--------------------
+
+``google.cloud.client._ClientFactoryMixin``
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+- Provides only ``from_service_account_json`` classmethod as a factory: it
+  loads credentials from a service account file and delegates to the main
+  class ctor with loaded credentials in ``**kwargs``.
+
+- If subclasses define ``_SET_PROJECT``, copies project ID from the
+  credentials into the ``**kwargs``. (This feature is only used by
+
+
+``google.cloud.client.Client``
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+- Mixes in ``_ClientFactoryMixin``.
+
+- Ctor forces ``credentials`` to be ``google.auth.credentials.Credentials``,
+  if passed.
+
+- Ctor takes optional ``_http`` (eewwww!), which is expected to be a
+  ``requests.Session`` workalike.
```

Comment:
Testing hack?

Reply:
I'm not sure at this point.  We've had requests to make the argument public, or otherwise allow app developers to provide a replacement, e.g. #4542.